### ISSUE TITLE
Callback on each transition of a good way to visualize the result of each step beyond text

### ISSUE THREAD

**Issue by ilykos** [pipefunc/pipefunc#393]
As I am working with functions that are often used outside of the pipeline (invoked directly), I would like to keep them as general as possible. In my own DAG implementation, I can create callbacks on every transition between nodes to visualize the output of each node using matplotlib. I would like to avoid inspecting files, as this may imply a significant serialization delay on the type of data I work with.

If I wanted to show the current progress of each step, what would be the most elegant way to create an implicit plotting routine between each transition?

**Comment by basnijholt** [comment]
Hi @ilykos, 

Would https://github.com/pipefunc/pipefunc/pull/306 work for you?

I think that is exactly what you are asking for.

Sorry it took me a month to reply 😅

**Comment by basnijholt** [comment]
Here's a full example:
```python
from pipefunc import Pipeline, pipefunc
import matplotlib.pyplot as plt
import numpy as np
from IPython.display import clear_output


def visualize_hook(func, result, kwargs):
    """Visualization hook that creates a plot for each function's output."""
    # Clear the current figure to update in-place in notebooks
    clear_output(wait=True)

    plt.figure(figsize=(10, 4))

    # Different visualization based on the function's output_name
    if func.output_name == "signal":
        plt.plot(result, label="Generated Signal")
        plt.title("Step 1: Generated Signal")
    elif func.output_name == "filtered":
        plt.plot(result, label="Filtered Signal", color="orange")
        plt.title("Step 2: Filtered Signal")
    elif func.output_name == "peaks":
        # Plot both the signal and detected peaks
        signal = kwargs["filtered"]
        plt.plot(signal, label="Signal", color="blue", alpha=0.5)
        plt.plot(result, signal[result], "ro", label="Peaks")
        plt.title("Step 3: Peak Detection")

    plt.grid(True)
    plt.legend()
    plt.xlabel("Sample")
    plt.ylabel("Amplitude")

    # Add progress indicator
    plt.suptitle(f"Pipeline Progress: Processing {func.output_name}", y=1.05)

    plt.tight_layout()
    plt.show()


# Define pipeline functions with visualization hooks
@pipefunc(output_name="signal", post_execution_hook=visualize_hook)
def generate_signal(n_samples: int, freq: float) -> np.ndarray:
    """Generate a noisy sinusoidal signal."""
    t = np.linspace(0, 10, n_samples)
    signal = np.sin(2 * np.pi * freq * t)
    noise = np.random.normal(0, 0.2, n_samples)
    return signal + noise


@pipefunc(output_name="filtered", post_execution_hook=visualize_hook)
def filter_signal(signal: np.ndarray, window: int = 5) -> np.ndarray:
    """Apply simple moving average filter."""
    return np.convolve(signal, np.ones(window) / window, mode="same")


@pipefunc(output_name="peaks", post_execution_hook=visualize_hook)
def find_peaks(filtered: np.ndarray, threshold: float = 0.5) -> np.ndarray:
    """Find peaks in the filtered signal."""
    peaks = []
    for i in range(1, len(filtered) - 1):
        if (
            filtered[i] > filtered[i - 1]
            and filtered[i] > filtered[i + 1]
            and filtered[i] > threshold
        ):
            peaks.append(i)
    return np.array(peaks)


# Create and run the pipeline
pipeline = Pipeline([generate_signal, filter_signal, find_peaks])
result = pipeline(n_samples=1000, freq=1.0, threshold=0.5)
```


I will push the PR https://github.com/pipefunc/pipefunc/pull/306 over the finishing line and release a new version right away!

**Comment by basnijholt** [comment]
@ilykos, I just released v0.41.0. Let me know what you think! 😄
